File: apkEditor/src/main/java/com/mcal/folderlist/FolderListWrapper.java
package com.mcal.folderlist;

import static com.mcal.common.utils.FileHelperKt.deleteAll;

import android.content.Context;
import android.text.InputFilter;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.AdapterView.OnItemLongClickListener;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.mcal.apkeditor.R;
import com.mcal.common.utils.InputHelper;
import com.mcal.folderlist.util.OpenFiles;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FolderListWrapper implements OnItemClickListener, OnItemLongClickListener {
    private final Context mContext;
    private final ListView mListView;
    private final String mRootPath;
    private final String mCurPath;

    private FolderListAdapter adapter;
    private final IListEventListener mListener;

    public FolderListWrapper(Context context, ListView listView, String curPath, String rootPath, IListEventListener listener, IListItemProducer producer) {
        mContext = context;
        mListView = listView;
        mCurPath = curPath;
        mRootPath = rootPath;
        mListener = listener;

        init(producer);
    }

    public FolderListAdapter getAdapter() {
        return adapter;
    }

    private void init(IListItemProducer producer) {
        adapter = new FolderListAdapter(mContext, mRootPath, mCurPath, producer);
        mListView.setAdapter(adapter);

        mListView.setOnItemClickListener(this);
        mListView.setOnItemLongClickListener(this);
    }

    @Override
    public void onItemClick(AdapterView<?> arg0, View arg1, int position,
                            long arg3) {
        final List<FileRecord> fileList = new ArrayList<>();
        final String oldDir = adapter.getData(fileList);
        final FileRecord rec = fileList.get(position);
        if (rec == null) {
            return;
        }

        if (rec.isDir) {
            final String targetPath;
            if (rec.fileName.equals("..")) {
                int pos = oldDir.lastIndexOf('/');
                targetPath = oldDir.substring(0, pos);
            } else {
                targetPath = oldDir + "/" + rec.fileName;
            }
            adapter.openDirectory(targetPath);
        } else {
            final String filePath = oldDir + "/" + rec.fileName;
            // When listener not deal with the opening, we will do it
            if (!mListener.fileClicked(arg1, filePath)) {
                OpenFiles.openFile(mContext, filePath);
            }
        }

        final String newDir = adapter.getData(null);
        if (!newDir.equals(oldDir)) {
            mListener.dirChanged(newDir);
        }
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, final int position, long id) {
        // The first item is always the parent folder
        if (position == 0) {
            return true;
        }
        parent.setOnCreateContextMenuListener((menu, v, menuInfo) -> {
            // Open As
            final MenuItem openAs = menu.add(0, Menu.FIRST, 0, "Open as...");
            openAs.setOnMenuItemClickListener(item -> {
                final List<FileRecord> fileList = new ArrayList<>();
                final String oldDir = adapter.getData(fileList);
                final FileRecord rec = fileList.get(position);
                if (rec != null) {
                    final String filePath = oldDir + "/" + rec.fileName;
                    OpenFiles.openFile(mContext, filePath);
                }
                return true;
            });

            // Delete
            final MenuItem item1 = menu.add(0, Menu.FIRST, 1, R.string.delete);
            item1.setOnMenuItemClickListener(item -> {
                deleteFile(position);
                return true;
            });

            // Rename
            final MenuItem item2 = menu.add(0, Menu.FIRST + 2, 0, R.string.rename);
            item2.setOnMenuItemClickListener(item -> {
                showRenameDlg(position);
                return true;
            });

            // New File
            final MenuItem item3 = menu.add(0, Menu.FIRST + 3, 0, R.string.new_file);
            item3.setOnMenuItemClickListener(item -> {
                createFile();
                return true;
            });

            if (mListener != null) {
                mListener.itemLongClicked(menu, v, menuInfo);
            }
        });
        return false;
    }

    private void showRenameDlg(int position) {
        final MaterialAlertDialogBuilder renameDlg = new MaterialAlertDialogBuilder(mContext);

        renameDlg.setTitle(R.string.rename);
        renameDlg.setMessage(R.string.pls_input_filename);

        // Set an EditText view to get user input
        final EditText input = new EditText(mContext);
        final List<FileRecord> records = new ArrayList<>();
        final String dirPath = adapter.getData(records);
        final FileRecord fr = records.get(position);
        if (fr == null) {
            return;
        }
        final String fileName = fr.fileName;
        input.setText(fileName);
        renameDlg.setView(input);

        renameDlg.setPositiveButton(android.R.string.ok,
                (dialog, whichButton) -> {
                    String newName = input.getText().toString();
                    boolean ret = doRename(dirPath, fileName, newName);
                    if (ret) {
                        adapter.fileRenamed(dirPath, fileName, newName);
                        mListener.fileRenamed(dirPath, fileName, newName);
                    }
                });

        renameDlg.setNegativeButton(android.R.string.cancel,
                (dialog, whichButton) -> {
                    // Canceled.
                });

        renameDlg.show();

    }

    protected boolean doRename(String dirPath, String fileName, String newName) {
        boolean ret = false;
        final File newFile = new File(dirPath + "/" + newName);
        if (newFile.exists()) {
            final String tip = mContext.getResources().getString(
                    R.string.file_already_exist);
            final String msg = String.format(tip, newName);
            Toast.makeText(mContext, msg, Toast.LENGTH_SHORT).show();
        } else {
            ret = new File(dirPath + "/" + fileName).renameTo(newFile);
            final String strRename = mContext.getResources().getString(R.string.rename);
            final String strResult = mContext.getResources().getString(
                    ret ? R.string.succeed : R.string.failed);

            Toast.makeText(mContext, strRename + " " + strResult, Toast.LENGTH_SHORT)
                    .show();
        }
        return ret;
    }

    private void createFile() {
        final String dirPath = adapter.getData(null);

        final MaterialAlertDialogBuilder inputDlg = new MaterialAlertDialogBuilder(mContext);
        inputDlg.setTitle(R.string.new_file);
        inputDlg.setMessage(R.string.pls_input_filename);

        // Set an EditText view to get user input
        final EditText input = new EditText(mContext);
        final InputFilter filter = InputHelper.getFileNameFilter();
        input.setFilters(new InputFilter[]{filter});
        inputDlg.setView(input);

        inputDlg.setPositiveButton(android.R.string.ok,
                (dialog, whichButton) -> {
                    String name = input.getText().toString();
                    name = name.trim();
                    if ("".equals(name)) {
                        Toast.makeText(mContext,
                                        R.string.empty_input_tip, Toast.LENGTH_LONG)
                                .show();
                    } else {
                        boolean succeed = false;
                        String errMessage = null;

                        // Try to create a new file in current directory
                        final File dir = new File(dirPath);
                        final File newFile = new File(dir, name);
                        try {
                            succeed = newFile.createNewFile();
                            if (succeed) {
                                // Update list view
                                adapter.openDirectory(dirPath);
                            } else {
                                errMessage = mContext.getString(R.string.failed_create_file);
                            }
                        } catch (IOException e) {
                            final String fmt = mContext.getString(R.string.general_error);
                            errMessage = String.format(fmt, e.getMessage());
                        }
                        if (!succeed) {
                            Toast.makeText(mContext, errMessage, Toast.LENGTH_LONG).show();
                        }
                    }
                });

        inputDlg.setNegativeButton(android.R.string.cancel,
                (dialog, whichButton) -> {
                    // Canceled.
                });

        inputDlg.show();
    }

    private void deleteFile(int position) {
        final List<FileRecord> records = new ArrayList<>();
        final String dirPath = adapter.getData(records);
        final FileRecord fr = records.get(position);
        if (fr == null) {
            return;
        }

        final String fileName = fr.fileName;
        final String path = dirPath + "/" + fr.fileName;
        final boolean ret = deleteFile(path);
        if (ret) {
            adapter.fileDeleted(dirPath, fileName);
            mListener.fileDeleted(dirPath, fileName);
        }
    }

    private boolean deleteFile(String path) {
        boolean ret = false;
        final File file = new File(path);

        if (file.exists()) {
            if (file.isFile()) {
                ret = file.delete();
            } else {
                try {
                    deleteAll(file);
                    ret = true;
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        return ret;
    }

    public void openDirectory(String dir) {
        final String oldDir = adapter.getData(null);

        // It may fail, so we need to check after the call
        adapter.openDirectory(dir);

        final String newDir = adapter.getData(null);

        if (!newDir.equals(oldDir)) {
            mListener.dirChanged(newDir);
        }
    }
}

File: apkEditor/src/main/java/com/mcal/folderlist/FolderListWrapper.kt
package com.mcal.folderlist

import android.content.Context
import android.content.DialogInterface
import android.widget.EditText
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.apkeditor.R
import com.mcal.common.utils.FileRecord
import com.mcal.common.utils.InputHelper
import com.mcal.common.utils.deleteAll
import com.mcal.folderlist.util.OpenFiles
import java.io.File
import java.io.IOException

open class FolderListWrapper(
    private val mContext: Context,
    private val mListView: RecyclerView,
    private val mCurPath: String,
    private val mRootPath: String,
    private val mListener: IListEventListener?,
    producer: IListItemProducer
) : FolderListAdapter.FileItemClick {
    var mAdapter: FolderListAdapter? = null
        private set

    private fun init(producer: IListItemProducer) {
        mAdapter = FolderListAdapter(this, mRootPath, mCurPath, producer)
        mListView.adapter = mAdapter
        mListView.layoutManager = LinearLayoutManager(mContext)
        mListView.adapter = mAdapter
    }

    override fun onClick(position: Int) {
        val fileList: MutableList<FileRecord> = ArrayList()
        val adapter = mAdapter
        val listener = mListener
        if (adapter != null && listener != null) {
            val oldDir = adapter.getData(fileList)
            val rec = fileList[position]
            if (rec.isDir && oldDir != null) {
                val targetPath: String = if (rec.fileName == "..") {
                    val pos = oldDir.lastIndexOf('/')
                    oldDir.substring(0, pos)
                } else {
                    oldDir + "/" + rec.fileName
                }
                adapter.openDirectory(targetPath)
            } else {
                val filePath = oldDir + "/" + rec.fileName
                // When listener not deal with the opening, we will do it
                if (!listener.fileClicked(filePath)) {
                    OpenFiles.openFile(mContext, filePath)
                }
            }
            val newDir = adapter.getData(null)
            if (newDir != oldDir) {
                listener.dirChanged(newDir)
            }
        }
    }

    override fun onLongClick(position: Int): Boolean {
        val context = mContext
        // The first item is always the parent folder
        if (position == 0) {
            return true
        }
        val dialog = MaterialAlertDialogBuilder(context)
        dialog.setItems(
            arrayOf(
                context.getString(R.string.file_open_as),
                context.getString(R.string.delete),
                context.getString(R.string.rename),
                context.getString(R.string.new_file)
            )
        ) { p112: DialogInterface, p2: Int ->
            when (p2) {
                0 -> {
                    val fileList: MutableList<FileRecord> = ArrayList()
                    val oldDir = mAdapter?.getData(fileList)
                    val rec = fileList[position]
                    if (rec != null) {
                        val filePath = oldDir + "/" + rec.fileName
                        OpenFiles.openFile(context, filePath)
                    }
                    p112.dismiss()
                }
                1 -> {
                    deleteFile(position)
                    p112.dismiss()
                }
                2 -> {
                    showRenameDlg(position)
                    p112.dismiss()
                }
                3 -> {
                    createFile()
                    p112.dismiss()
                }
            }
        }
        dialog.create().show()
        mListener?.itemLongClicked()
        return false
    }

    private fun showRenameDlg(position: Int) {
        val renameDlg = MaterialAlertDialogBuilder(mContext)
        renameDlg.setTitle(R.string.rename)
        renameDlg.setMessage(R.string.pls_input_filename)

        // Set an EditText view to get user input
        val input = EditText(mContext)
        val records: MutableList<FileRecord> = ArrayList()
        val fr = records[position]
        val fileName = fr.fileName
        input.setText(fileName)
        renameDlg.setView(input)
        renameDlg.setPositiveButton(android.R.string.ok) { _, _ ->
            val newName = input.text.toString()
            val adapter = mAdapter
            if (adapter != null) {
                val dirPath = adapter.getData(records)
                if (dirPath != null && fileName != null) {
                    val ret = doRename(dirPath, fileName, newName)
                    if (ret) {
                        adapter.fileRenamed(dirPath, fileName, newName)
                        adapter.fileRenamed(dirPath, fileName, newName)
                    }
                }
            }
        }
        renameDlg.setNegativeButton(android.R.string.cancel, null)
        renameDlg.show()
    }

    private fun doRename(dirPath: String, fileName: String, newName: String): Boolean {
        val context = mContext
        var ret = false
        val newFile = File("$dirPath/$newName")
        if (newFile.exists()) {
            val tip = context.resources.getString(R.string.file_already_exist)
            val msg = String.format(tip, newName)
            Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
        } else {
            ret = File("$dirPath/$fileName").renameTo(newFile)
            val strRename = context.resources.getString(R.string.rename)
            val strResult = context.resources.getString(if (ret) R.string.succeed else R.string.failed)
            Toast.makeText(context, "$strRename $strResult", Toast.LENGTH_SHORT).show()
        }
        return ret
    }

    private fun createFile() {
        val context = mContext
        val dirPath = mAdapter?.getData(null)
        val inputDlg = MaterialAlertDialogBuilder(context)
        inputDlg.setTitle(R.string.new_file)
        inputDlg.setMessage(R.string.pls_input_filename)

        // Set an EditText view to get user input
        val input = EditText(context)
        val filter = InputHelper.getFileNameFilter()
        input.filters = arrayOf(filter)
        inputDlg.setView(input)
        inputDlg.setPositiveButton(android.R.string.ok) { _, _ ->
            var name = input.text.toString()
            name = name.trim { it <= ' ' }
            if ("" == name) {
                Toast.makeText(context, R.string.empty_input_tip, Toast.LENGTH_LONG).show()
            } else {
                var succeed = false
                var errMessage: String? = null

                // Try to create a new file in current directory
                val dir = File(dirPath)
                val newFile = File(dir, name)
                try {
                    succeed = newFile.createNewFile()
                    if (succeed) {
                        // Update list view
                        mAdapter?.openDirectory(dirPath)
                    } else {
                        errMessage = context.getString(R.string.failed_create_file)
                    }
                } catch (e: IOException) {
                    val fmt = context.getString(R.string.general_error)
                    errMessage = String.format(fmt, e.message)
                }
                if (!succeed) {
                    Toast.makeText(context, errMessage, Toast.LENGTH_LONG).show()
                }
            }
        }
        inputDlg.setNegativeButton(android.R.string.cancel, null)
        inputDlg.show()
    }

    private fun deleteFile(position: Int) {
        val records: MutableList<FileRecord> = ArrayList()
        mAdapter?.let { adapter ->
            val dirPath = adapter.getData(records)
            val fr = records[position]
            val fileName = fr.fileName
            val path = dirPath + "/" + fr.fileName
            val ret = deleteFile(path)
            if (ret && fileName != null) {
                adapter.fileDeleted(dirPath, fileName)
                mListener?.fileDeleted(dirPath, fileName)
            }
        }
    }

    private fun deleteFile(path: String): Boolean {
        var ret = false
        val file = File(path)
        if (file.exists()) {
            if (file.isFile) {
                ret = file.delete()
            } else {
                try {
                    deleteAll(file)
                    ret = true
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        return ret
    }

    fun openDirectory(dir: String?) {
        mAdapter?.let { adapter ->
            val oldDir = adapter.getData(null)

            // It may fail, so we need to check after the call
            adapter.openDirectory(dir)
            val newDir = adapter.getData(null)
            if (newDir != oldDir) {
                mListener?.dirChanged(newDir)
            }
        }
    }

    init {
        init(producer)
    }
}